### PR TITLE
MULE-13009: Inconsistent error propagation in extension operations depending on operation type

### PR DESCRIPTION
**Pull request by juandesi**

### CODE REVIEW

File: core/src/main/java/org/mule/runtime/core/api/exception/NullExceptionHandler.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import org.mule.runtime.core.api.processor.MessageProcessorChain;
+import org.reactivestreams.Publisher;
+import reactor.core.publisher.Mono;
+
+/**
+ * Null {@link MessagingExceptionHandler} which can be used to configure a {@link MessageProcessorChain} to not handle errors.
+ *
+ * @since 4.0
+ */
+public class NullExceptionHandler implements MessagingExceptionHandler {
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/api/exception/NullExceptionHandler.java]: final

File: core/src/main/java/org/mule/runtime/core/api/util/MessagingExceptionResolver.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+   * This method will pick the FIRST cause exception that has a mule or extension KNOWN error as the real cause, if there is not
+   * an exception in the causes that match with an Known error type then this method will try to find the error that the current
+   * {@link Event} is carrying.
+   * <p>
+   * When there are multiple exceptions that contains the same root error type, then this method will wrap the one that has
+   * highest position in the causes list
+   *
+   * @return a {@link MessagingException} with the proper {@link Error} associated to it's {@link InternalEvent}
+   */
+  public static MessagingException resolve(Processor processor, MessagingException me, MuleContext context) {
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/api/util/MessagingExceptionResolver.java]: having this as a static method might be convenient, but prevents this from being mock and build tests in which a specific exception is resolved. 

Consider turning this into an instance method. If you have a good reason not to, then this class needs to be final and have a private constructor

File: core/src/main/java/org/mule/runtime/core/api/util/MessagingExceptionResolver.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Tries to resolve a new {@link MessagingException}s with the real cause of the problem from an incoming
+ * {@link MessagingException} that carries a list of causes inside it so it can be thrown and no information is lost.
+ *
+ * @since 4.0
+ */
+public class MessagingExceptionResolver {
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/api/util/MessagingExceptionResolver.java]: what if an extension needs to use this? Shouldn't this be mule api?

Comment (juandesi) [inline core/src/main/java/org/mule/runtime/core/api/util/MessagingExceptionResolver.java]: no way, this is 100% concern of the runtime

File: core/src/main/java/org/mule/runtime/core/api/util/ExceptionUtils.java [hunk 4]
```diff
@@ -212,246 +147,114 @@ public static String getFullStackTraceWithoutMessages(Throwable throwable) {
...
-        .filter(error -> errorCauseMatchesException(causeException, error)
-            || messagingException.causedExactlyBy(error.getCause().getClass()))
-        .isPresent()) {
-      Error newError = getErrorFromFailingProcessor(annotatedObject, causeException, errorTypeLocator);
-      InternalEvent event = InternalEvent.builder(messagingException.getEvent()).error(newError).build();
-      messagingException.setProcessedEvent(event);
-      return event;
-    } else {
-      return currentEvent;
+  static List<ErrorMapping> getErrorMappings(Object processor) {
```
Comment (afelisatti) [inline core/src/main/java/org/mule/runtime/core/api/util/ExceptionUtils.java]: protected?

File: core/src/main/java/org/mule/runtime/core/api/util/ExceptionUtils.java [hunk 5]
```diff
@@ -212,246 +147,114 @@ public static String getFullStackTraceWithoutMessages(Throwable throwable) {
...
-    if (failing == null) {
-      failing = processor;
-      exception = new MessagingException(createStaticMessage(rootException.getMessage()), exception.getEvent(),
-                                         rootException instanceof MessagingException ? rootException.getCause() : rootException,
-                                         processor);
-    }
-    exception.setProcessedEvent(createErrorEvent(exception.getEvent(), processor, exception, errorTypeLocator));
-    return putContext(exception, failing, exception.getEvent(), muleContext);
+  static boolean isUnknownMuleError(ErrorType type) {
+    return type.getIdentifier().equals(UNKNOWN_ERROR_IDENTIFIER);
```
Comment (afelisatti) [inline core/src/main/java/org/mule/runtime/core/api/util/ExceptionUtils.java]: Consider namespace as well.

File: core/src/main/java/org/mule/runtime/core/api/util/ExceptionUtils.java [hunk 6]
```diff
@@ -212,246 +147,114 @@ public static String getFullStackTraceWithoutMessages(Throwable throwable) {
...
-    }
-
-    if (errorMappings != null && !errorMappings.isEmpty()) {
-      Optional<ErrorMapping> matchedErrorMapping = errorMappings.stream().filter(mapping -> mapping.match(errorType)).findFirst();
-      if (matchedErrorMapping.isPresent()) {
-        return matchedErrorMapping.get().getTarget();
-      }
-    }
-    return errorType;
+  private static ErrorType getErrorTypeFromFailingProcessor(Object processor, Throwable exception,
```
Comment (afelisatti) [inline core/src/main/java/org/mule/runtime/core/api/util/ExceptionUtils.java]: Do these have javadocs? hard to see in the diff...

File: core/src/main/java/org/mule/runtime/core/api/util/MessagingExceptionResolver.java [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+
+  /**
+   * Resolves a new {@link MessagingException} with the real cause of the problem based on the content of an
+   * Incoming {@link MessagingException} with a chain of causes inside it and the current event that the exception is carrying.
```
Comment (afelisatti) [inline core/src/main/java/org/mule/runtime/core/api/util/MessagingExceptionResolver.java]: incoming

File: core/src/main/java/org/mule/runtime/core/api/util/MessagingExceptionResolver.java [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+
+    if (!rootCause.isPresent()) {
+      return updateCurrent(me, processor, context);
+    }
+
+    Throwable root = rootCause.get().getFirst();
+    ErrorType rootErrorType = rootCause.get().getSecond();
+    Processor failingProcessor = getFailingProcessor(me, root).orElse(processor);
+
+    ErrorType errorType = getErrorMappings(processor)
```
Comment (afelisatti) [inline core/src/main/java/org/mule/runtime/core/api/util/MessagingExceptionResolver.java]: Why are mappings applied twice? Here and in `ExceptionUtils`.

File: core/src/main/java/org/mule/runtime/core/internal/exception/ErrorTypeLocatorFactory.java [hunk 9]
```diff
@@ -85,6 +85,7 @@ public static ErrorTypeLocator createDefaultErrorTypeLocator(ErrorTypeRepository
                                  errorTypeRepository.lookupErrorType(STREAM_MAXIMUM_SIZE_EXCEEDED).get())
             .addExceptionMapping(MuleFatalException.class, errorTypeRepository.getErrorType(FATAL).get())
             .addExceptionMapping(CompositeRoutingException.class, errorTypeRepository.getErrorType(COMPOSITE_ROUTING).get())
+            .addExceptionMapping(LinkageError.class, errorTypeRepository.getErrorType(FATAL).get())
```
Comment (afelisatti) [inline core/src/main/java/org/mule/runtime/core/internal/exception/ErrorTypeLocatorFactory.java]: Why was this necessary? AFAIK that error should be wrapped to a MuleFatalException.

**Comment by afelisatti** [comment]
Also, make sure you have tests cases for the UNKNOWN case, when there's nothing but that to be found.

File: core/src/main/java/org/mule/runtime/core/api/util/ExceptionUtils.java (hunk 6 above)
Comment (juandesi) [inline core/src/main/java/org/mule/runtime/core/api/util/ExceptionUtils.java]: nope, those are trivial private methods

**Comment by afelisatti** [comment]
@juandesi just let me know when the rebase is done and I'll approve it again.
